packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -254,6 +254,13 @@ export class Outbox {
 		// Don't use resubmittingBatchId for idAllocationBatch.
 		// ID Allocation messages are not directly resubmitted so we don't want to reuse the batch ID.
 		this.flushInternal(this.idAllocationBatch);
+		// We need to flush an empty batch if the main batch *becomes* empty on resubmission.
+		// The way to catch this is to check if both blobAttachBatch and mainBatch are empty. This is because
+		// blob attach batches can't *become* empty on resubmission, if it's empty it was already empty before.
+		if (resubmittingBatchId && this.blobAttachBatch.empty && this.mainBatch.empty) {
```

Comment:
ohhhh nice!!  You cracked the code!  Haha but really, this is an elegant / crystal clear way to implement this.